source/Halibut/ServiceModel/PendingRequestQueueAsync.cs

```diff
@@ -0,0 +1,250 @@
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Halibut.Diagnostics;
+using Halibut.Transport.Protocol;
+using Nito.AsyncEx;
+
+namespace Halibut.ServiceModel
+{
+    public class PendingRequestQueueAsync : IPendingRequestQueue
+    {
+        readonly ConcurrentQueue<PendingRequest> queue = new();
+        readonly Dictionary<string, PendingRequest> inProgress = new();
+        readonly SemaphoreSlim queueLock = new(1, 1);
+        readonly AsyncManualResetEvent itemAddedToQueue = new(false);
+        readonly ILog log;
+        readonly TimeSpan pollingQueueWaitTimeout;
+
+        public PendingRequestQueueAsync(ILog log) : this(log, HalibutLimits.PollingQueueWaitTimeout)
+        {
+            this.log = log;
+        }
+
+        public PendingRequestQueueAsync(ILog log, TimeSpan pollingQueueWaitTimeout)
+        {
+            this.log = log;
+            this.pollingQueueWaitTimeout = pollingQueueWaitTimeout;
+        }
+        
+        public async Task<ResponseMessage> QueueAndWaitAsync(RequestMessage request, CancellationToken cancellationToken)
+        {
+            var pending = new PendingRequest(request, log);
+
+            using (await queueLock.LockAsync(cancellationToken))
```

Comment:
As per [Luke's question](https://octopusdeploy.slack.com/archives/C04JFJXECS1/p1690869232305179), if we decide we should, then I should retro `ConfigureAwait` in this PR...